llvm/lib/Support/Parallel.cpp

```diff
@@ -217,13 +194,18 @@ TaskGroup::~TaskGroup() {
 void TaskGroup::spawn(std::function<void()> F, bool Sequential) {
 #if LLVM_ENABLE_THREADS
   if (Parallel) {
+    if (Sequential) {
+      // Act as worker thread 0.
+      threadIndex = 0;
```

Comment:
Since this sequential behavior was only added (and seems only to be used by) the ELF relocations, can we not move this code there?

Reply:
Could you elaborate the expected code here? `threadIndex` is to satisfy lld ELF relocation's requirement. It does need `threadIndex == 0`. The initial `threadIndex` is UINT_MAX is for assertion https://reviews.llvm.org/D148916

Reply:
IIUC, if a user was to mix sequential and parallel tasks within the same `TaskGroup` then this could potentially lead to unintentional concurrent access to thread index `0` resources.

Reply:
@MaskRay I meant this whole sequential flag could be solved by including the for... loop in the task at the call site, when serial=true, no? (in elf::scanRelocations) That way you get a valid `threadIndex` and the items are executed sequentially.
```
  bool serial = ...
  parallel::TaskGroup tg;
  auto outerFn = [&]() {
    for (ELFFileBase *f : ctx.objectFiles) {
      auto fn = [f]() {
        RelocationScanner scanner;
        for (InputSectionBase *s : f->getSections()) {
          if (s && s->kind() == SectionBase::Regular && s->isLive() &&
              (s->flags & SHF_ALLOC) &&
              !(s->type == SHT_ARM_EXIDX && config->emachine == EM_ARM))
            scanner.template scanSection<ELFT>(*s);
        }
      };
      if (serial)
        fn();
      else
        tg.spawn(fn);

    }
  };
  if (serial)
    tg.spawn(outerFn);
  else
    outerFn();
```

Reply:
Thus we don't need the `Sequential` flag here in Parallel.cpp

Reply:
I think this suggestion is what I previously did (see the deleted code on https://reviews.llvm.org/D148728)
```cpp
    if (serial)
      fn();
    else
      tg.execute(fn);
```

It seems convenient to have a Parallel helper function to simplify this pattern. I do agree that the load-bearing `spawn` should not have extra overhead (`bool Sequential = false`) but that is not the focus of this patch.

---

Perhaps the next patch can add:

```
spawn(std::function f)
execute(std::function f, bool serial)
```

Reply:
If the threadIndex would be set to 0 then it might clash with another task which is already has threadIndex equal 0. In that cast the resource with index 0 would be accessed concurrently which is unwanted behavior.

if we will use old behavior:

`if (serial) 
    fn(); 
  else 
tg.execute(fn);
`

then we will have assertion checking that thread_index is not equal UINT_MAX. (thread_index is equal to UINT_MAX for main thread)

In my patch I use separate thread for sequential tasks. It looks like this solve the problem when thread_index equal UINT_MAX and when zero index clashes with another thread.

Reply:
This patch is an almost revert of reviews.llvm.org/D148728 . 

Since the use case is so special, I'm fine with the style suggested by @aganea (also the initial style I used for parallel relocation scanning):

```cpp
      if (serial)
        fn();
      else
        tg.spawn(fn);

    }
  };
  if (serial)
    tg.spawn(outerFn);
  else
    outerFn();
```

Perhaps Parallel.cpp could add `setThreadIndex` and lld/ELF/Relocations.cpp could call `setThreadIndex(0)` to make the pattern more explicit. It's up to lld to ensure correctness.

Reply:
if this code would be left in the patch:

```
    if (Sequential) {
      // Act as worker thread 0.
      threadIndex = 0;   <<<<<<<<<<<<<<<<<<<
```

then threadIndex = 0 could clash with another thread having threadIndex = 0. Which might result in concurrent access to the resource with index 0.

Reply:
@avl-llvm Yes, I think the goal would be to remove this chunk of code (everything related to `Sequential`)

Reply:
I see, thanks. So removing Sequential mode and including the for... loop in the task at the call site in elf::scanRelocations would OK, then.

Reply:
Thanks. I just realized the flipped condition :) Adopted the style and kept the `threadIndex != UINT_MAX` assertion.